### PR TITLE
Add some variance tests

### PR DESCRIPTION
**Pull request by elliottt**
### Motivation

Add some tests of generic variance to show what functionality exists, and what functionality is needed.

### Test plan

See included automated tests.

### CODE REVIEW

File: test/testdata/infer/generics/disabled/variance_methods.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class A
+  extend T::Sig
+  extend T::Helpers
+  extend T::Generic
+
+  abstract!
+
+  Key = type_member(:in) # error: Classes can only have invariant type members
+  Val = type_member(:out)
```
Comment (DarkDimius) [inline test/testdata/infer/generics/disabled/variance_methods.rb]: I expected this to be an error too

File: test/testdata/infer/generics/disabled/variance_methods.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# typed: true
+
+class A
+  extend T::Sig
+  extend T::Helpers
+  extend T::Generic
+
+  abstract!
+
+  Key = type_member(:in) # error: Classes can only have invariant type members
```
Comment (DarkDimius) [inline test/testdata/infer/generics/disabled/variance_methods.rb]: Maybe call type params `Co` and `Contra` to make it easier to read?

File: test/testdata/infer/generics/disabled/variance_methods.rb (hunk 1 above)
Comment (elliottt) [reply test/testdata/infer/generics/disabled/variance_methods.rb]: Yes, me too. Oddly it only raises an error for the first one.

File: test/testdata/infer/generics/disabled/variance_params.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# typed: true
```
Comment (DarkDimius) [inline test/testdata/infer/generics/disabled/variance_params.rb]: why is this test disabled? What part of it fails?

File: test/testdata/infer/generics/disabled/variance_params.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+class Cat < Animal; end
+
+class Other; end
+
+class Makes
+  extend T::Sig
+  extend T::Helpers
+  extend T::Generic
+
+  Ty = type_member(:out) # error: Classes can only have invariant type members
```
Comment (DarkDimius) [inline test/testdata/infer/generics/disabled/variance_params.rb]: could you make these into modules?

File: test/testdata/infer/generics/disabled/variance_params.rb (hunk 3 above)
Comment (elliottt) [reply test/testdata/infer/generics/disabled/variance_params.rb]: It only fails in LSP, as the fast-path doesn't raise the error about non-invariant type members.

Comment (DarkDimius) [reply test/testdata/infer/generics/disabled/variance_params.rb]: Can we leave a comment on this?

Comment (DarkDimius) [reply test/testdata/infer/generics/disabled/variance_params.rb]: (or make classes into modules)

File: test/testdata/infer/generics/disabled/variance_params.rb (hunk 4 above)
Comment (elliottt) [reply test/testdata/infer/generics/disabled/variance_params.rb]: Done. Thanks for the suggestion, I've re-enabled these tests as this passes in LSP now.

File: test/testdata/infer/generics/disabled/variance_params.rb (hunk 3 above)
Comment (elliottt) [reply test/testdata/infer/generics/disabled/variance_params.rb]: That worked, thanks!

File: test/testdata/infer/generics/disabled/variance_methods.rb (hunk 2 above)
Comment (elliottt) [reply test/testdata/infer/generics/disabled/variance_methods.rb]: Done, much easier to read!

File: test/testdata/infer/generics/variance_methods.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  In = type_member(:in)
+  Out = type_member(:out)
+
+  # should pass: the parameters are used correctly
+  sig {abstract.params(key: In).returns(Out)}
+  def lookup(key); end
+
+  # should pass: the parameters are used correctly, and the block argument swaps
+  # the polarity of the type members
+  sig {abstract.params(key: In, blk: T.proc.params(arg: Out).returns(In)).returns(Out)}
```
Comment (DarkDimius) [inline test/testdata/infer/generics/variance_methods.rb]: Do you want to try going through a variant generic class other that Proc?

Comment (elliottt) [reply test/testdata/infer/generics/variance_methods.rb]: I've added the `variance_user.rb` test to test variance with a user-defined type.

**Pull request merged**